plugins/kn-migration/cmd/migrate.go

```diff
@@ -0,0 +1,290 @@
+// Copyright © 2019 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"github.com/fatih/color"
+	"github.com/spf13/cobra"
+	apiv1 "k8s.io/api/core/v1"
+	api_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc" // from https://github.com/kubernetes/client-go/issues/345
+	"k8s.io/client-go/tools/clientcmd"
+	serving_v1_client "knative.dev/serving/pkg/client/clientset/versioned/typed/serving/v1"
+	"os"
+)
+
+// migrateCmd represents the migrate command
+var migrateCmd = &cobra.Command{
+	Use:   "migrate",
+	Short: "Migrate Knative services from source cluster to destination cluster",
+	Example: `
+  # Migrate Knative services from source cluster to destination cluster by export KUBECONFIG and KUBECONFIG_DESTINATION as environment variables
+  kn migrate --namespace default --destination-namespace default
+  # Migrate Knative services from source cluster to destination cluster by set kubeconfig as parameters
+  kn migrate --namespace default --destination-namespace default --kubeconfig $HOME/.kube/config/source-cluster-config.yml --destination-kubeconfig $HOME/.kube/config/destination-cluster-config.yml
+  # Migrate Knative services from source cluster to destination cluster and force replace the service if exists in destination cluster
+  kn migrate --namespace default --destination-namespace default --force
+  # Migrate Knative services from source cluster to destination cluster and delete the service in source cluster
+  kn migrate --namespace default --destination-namespace default --force --delete`,
+
+	Run: func(cmd *cobra.Command, args []string) {
```

Comment:
This function is WAY WAY too long. Divide it into smaller private functions with descriptive names. Further sub divide private functions as needed. No single functions should be longer than a page or terminal screen.

Reply:
separated to several small functions in the second commit: https://github.com/knative/client-contrib/pull/6/commits/e77b98b6cebdccc7a81fa81f9c508488611a2d3b